public/app/features/search/components/DashboardSearch.tsx

```diff
@@ -0,0 +1,212 @@
+import React, { FC, useReducer, useState } from 'react';
+import { useDebounce } from 'react-use';
+import { css } from 'emotion';
+import { Icon, useTheme, CustomScrollbar } from '@grafana/ui';
+import { getLocationSrv } from '@grafana/runtime';
+import { GrafanaTheme } from '@grafana/data';
+import { SearchSrv } from 'app/core/services/search_srv';
+import { backendSrv } from 'app/core/services/backend_srv';
+import { SearchQuery } from 'app/core/components/search/search';
+import { TagFilter } from 'app/core/components/TagFilter/TagFilter';
+import { contextSrv } from 'app/core/services/context_srv';
+import { DashboardSearchItemType, DashboardSection, OpenSearchParams } from '../types';
+import { findSelected, hasId, parseQuery } from '../utils';
+import { searchReducer, initialState } from '../reducers/dashboardSearch';
+import { getDashboardSrv } from '../../dashboard/services/DashboardSrv';
+import {
+  FETCH_ITEMS,
+  FETCH_RESULTS,
+  TOGGLE_SECTION,
+  MOVE_SELECTION_DOWN,
+  MOVE_SELECTION_UP,
+} from '../reducers/actionTypes';
+import { SearchField } from './SearchField';
+import { SearchResults } from './SearchResults';
+
+const searchSrv = new SearchSrv();
+
+const defaultQuery: SearchQuery = { query: '', parsedQuery: { text: '' }, tags: [], starred: false };
+const { isEditor, hasEditPermissionInFolders } = contextSrv;
+const canEdit = isEditor || hasEditPermissionInFolders;
+
+export interface Props {
+  closeSearch: () => void;
+  payload?: OpenSearchParams;
+}
+
+export const DashboardSearch: FC<Props> = ({ closeSearch, payload = {} }) => {
+  const [query, setQuery] = useState({ ...defaultQuery, ...payload, parsedQuery: parseQuery(payload.query) });
+  const [{ results, loading }, dispatch] = useReducer(searchReducer, initialState);
+  const theme = useTheme();
+  const styles = getStyles(theme);
+
+  useDebounce(
+    () => {
+      search();
+    },
+    300,
+    [query]
+  );
+
+  const search = () => {
+    let folderIds: number[] = [];
+    if (query.parsedQuery.folder === 'current') {
+      const { folderId } = getDashboardSrv().getCurrent().meta;
+      if (folderId) {
+        folderIds.push(folderId);
+      }
+    }
+    searchSrv.search({ ...query, tag: query.tags, query: query.parsedQuery.text, folderIds }).then(results => {
+      dispatch({ type: FETCH_RESULTS, payload: results });
+    });
+  };
+
+  const toggleSection = (section: DashboardSection) => {
+    if (hasId(section.title) && !section.items.length) {
+      backendSrv.search({ ...defaultQuery, folderIds: [section.id] }).then(items => {
+        dispatch({ type: FETCH_ITEMS, payload: { section, items } });
+        dispatch({ type: TOGGLE_SECTION, payload: section });
+      });
+    } else {
+      dispatch({ type: TOGGLE_SECTION, payload: section });
+    }
+  };
+
+  const onQueryChange = (searchQuery: string) => {
+    setQuery(q => ({
+      ...q,
+      parsedQuery: parseQuery(searchQuery),
```

Comment:
this is a prime example of why I find big large FC hard to read, mixing all these event handlers in the same render function, before we have a separation between markup rendering and handlers. Is there a way to convert this to a class component or move all the state handling and callbacks outside? All these functions are redefined on every render as well.

Reply:
yup, those should be either moved outside or wrapped on useCallback

Reply:
Not sure I understand 100%, this is tied directly to this component's state, so I don't see an easy way to just move this out of the component as it needs to set its state. Creating a wrapper component and moving all the logic there seems a bit excessive imo, also would create extra nesting. 

Refactoring this to class component we'd lose the possibility to use custom hooks (I was thinking that the whole searching action could be made into custom hook and shared with manage dashboards component), as well as handling state with use reducer. 

What I could do is to wrap the functions in `useCallback` to prevent their re-creation every render. When profiling this component that wasn't an issue and generally I'm not sure preemptively optimising everything would bring large benefits, but in this case, especially considering it can handle thousands of components, this seems like a good idea. Also depending how manage dashboards would end up looking, query handling could also be moved to a reducer.

Reply:
Moving things out of the render function is one big advantage of hooks, that way you can share state logic between components and test them in isolation

Reply:
yea, no need for useCallback  as this is the main outer component and is not re-rendered that frequently (except when filters / query change). So maybe skip that, useCallback  makes FC even worse in readability

Reply:
In case of function components, the whole component is a render function :D 

I'll try to refactor this logic into a custom hook when working on manage dashboards migration.

Reply:
https://github.com/grafana/grafana/blob/1ed23fcbb5f63afcae7ebb500fca7a3024c517c4/public/app/features/explore/TraceView/useChildrenState.ts